File: ts/project/.sdk/tm/c/utility/transform_response.c
// transform_response utility (mirrors utility/transform_response.rs).

#include "sdk.h"

voxgig_value* transform_response_util(Context* ctx) {
  Spec* spec = ctx->spec;
  SdkResult* result = ctx->result;
  voxgig_value* point = ctx->point;

  if (spec) spec_set_step(spec, "resform");

  if (!result) return voxgig_new_undef();
  if (!result->ok) return voxgig_new_undef();

  voxgig_value* transform = to_map(getp(point, "transform"));
  if (v_is_noval(transform)) return voxgig_new_undef();

  voxgig_value* resform = getp(transform, "res");
  if (v_is_noval(resform) || v_is_null(resform)) return voxgig_new_undef();

  voxgig_value* store = cmap(7,
      "ok", v_bool(result->ok),
      "status", v_num((double)result->status),
      "statusText", v_str(result->status_text),
      "headers", v_share(result->headers),
      "body", v_share(result->body),
      "resdata", v_share(result->resdata),
      "resmatch", v_share(result->resmatch));

  voxgig_value* out = voxgig_transform(store, resform, NULL);
  if (out && !v_is_noval(out)) {
    result->resdata = out;
    return out;
  }
  return voxgig_new_undef();
}

File: ts/project/.sdk/tm/rust/utility/transform_response.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::{getp, to_map};
use crate::utility::voxgigstruct as vs;
use crate::utility::voxgigstruct::Value;

pub fn transform_response_util(ctx: &Rc<Context>) -> Value {
    let spec = ctx.spec.borrow().clone();
    let result = ctx.result.borrow().clone();
    let point = ctx.point.borrow().clone();

    if let Some(sp) = &spec {
        sp.borrow_mut().step = "resform".to_string();
    }

    let result = match result {
        Some(r) => r,
        None => return Value::Noval,
    };
    if !result.borrow().ok {
        return Value::Noval;
    }

    let transform = to_map(&getp(&point, "transform"));
    if transform.is_noval() {
        return Value::Noval;
    }

    let resform = getp(&transform, "res");
    if resform.is_noval() || resform.is_null() {
        return Value::Noval;
    }

    let store = {
        let r = result.borrow();
        crate::core::helpers::jo(vec![
            ("ok", Value::Bool(r.ok)),
            ("status", Value::Num(r.status as f64)),
            ("statusText", Value::str(r.status_text.clone())),
            ("headers", r.headers.clone()),
            ("body", r.body.clone()),
            ("resdata", r.resdata.clone()),
            ("resmatch", r.resmatch.clone()),
        ])
    };

    match vs::transform(&store, &resform, None) {
        Ok(resdata) => {
            result.borrow_mut().resdata = resdata.clone();
            resdata
        }
        Err(_) => Value::Noval,
    }
}
